numpy/core/src/multiarray/dtype_transfer.c

```diff
@@ -161,6 +164,215 @@ _strided_to_strided_copy_references(char *dst, npy_intp dst_stride,
     return 0;
 }
 
+/************************** ANY TO OBJECT *********************************/
+
+typedef struct {
+    NpyAuxData base;
+    PyArray_GetItemFunc *getitem;
+    PyArrayObject_fields arr_fields;
+    PyArray_StridedUnaryOp *decref_func;
+    NpyAuxData *decref_data;
+} _any_to_object_auxdata;
+
+
+static void
+_any_to_object_auxdata_free(NpyAuxData *auxdata)
+{
+    _any_to_object_auxdata *data = (_any_to_object_auxdata *)auxdata;
+
+    Py_DECREF(data->arr_fields.descr);
+    NPY_AUXDATA_FREE(data->decref_data);
+    PyMem_Free(data);
+}
+
+
+static NpyAuxData *
+_any_to_object_auxdata_clone(NpyAuxData *auxdata)
+{
+    _any_to_object_auxdata *data = (_any_to_object_auxdata *)auxdata;
+
+    _any_to_object_auxdata *res = PyMem_Malloc(sizeof(_any_to_object_auxdata));
+
+    memcpy(res, data, sizeof(*data));
+    Py_INCREF(res->arr_fields.descr);
+    res->decref_data = NPY_AUXDATA_CLONE(data->decref_data);
+    if (res->decref_data == NULL) {
+        NPY_AUXDATA_FREE((NpyAuxData *)res);
+        return NULL;
+    }
+    return (NpyAuxData *)res;
```

Comment:
Is this covered by a test? Maybe worth trying to think of one that has a use for this code path?

Reply:
Yeah, the only way to reach this is using `np.nditer(...).copy()` probably, but can add tests.

Anything else that jumps at you aside from a few additional tests?  Most of the error paths (aside those in `array_method.c`) are probably (almost) impossible to tests, though.

Reply:
No, it all looks quite straightforward although repetitive. It would be nice to try to refactor this somehow, but I don't see how. There is not much that needs careful refcount considerations or malloc/free checking, since it is mostly logic to assign function pointers.

One thing to think about would be to split the large files into smaller pieces.